pkg/pulumiyaml/semver.go

```diff
@@ -7,15 +7,20 @@ import (
 	"github.com/pulumi/pulumi-yaml/pkg/pulumiyaml/ast"
 )
 
-func ParseVersion(v *ast.StringExpr) (*semver.Version, error) {
-	if v == nil || v.Value == "" {
-		return nil, nil
-	}
+func ParseVersion(expr ast.Expr) (*semver.Version, error) {
```

Comment:
Is this even used anymore? If we're allowing version to be any expression it probably isn't safe to have a function laying around that looks like it correctly gives a *semver.Version from an Expr.

Reply:
This is used in a few places where there isn't an evaluator available to evaluate the expression, so it tries to get the value of the version when it is statically known (i.e. `*ast.StringExpr`). 

>  it probably isn't safe to have a function laying around that looks like it correctly gives a *semver.Version from an Expr

The function isn't total, it will try to return a `*semver.Version` when it can from a string expression, otherwise it will return `nil` which the rest of the caller code handles already (when resolving resource/invoke versions) 

Maybe adding a `bool` to the returned tuple could make the signature a bit more clear. Returning `false` when the `Expr` is anything but `*ast.StringExpr`

Reply:
Are we allowing `version` to be any expression, or just a non-literal? Specifically, are we allowing `version` to be a outputty value?

I would strongly recommend that `version` must be knowable at preview time.

Reply:
Syntactically it can be any expression but the evaluation should return `string` at runtime. We check that when we run `evaluateExpr`:
```go
if versionExpr, ok := e.evaluateExpr(t.CallOpts.Version); ok {
	if version, ok := versionExpr.(string); ok {
		opts = append(opts, pulumi.Version(version))
	} else {
		e.error(t.CallOpts.Version, "version must be a string value")
	}
}
```

Reply:
Let's make sure that there is an issue to update the documentation on this issue then.